### DISCUSSION TITLE
sending unlock account email if account is locked

### DISCUSSION THREAD

**Discussion by rida-elbahtouri** [jeremyevans/rodauth#383] [General]
is there a way to send an unlock account email if the account is locked?

I tried this
```ruby
locked_out? do
      send_unlock_account_email
end
```

but I get an email with an empty key (key?=id_)
when doing something like this
```ruby
RodauthMailer.unlock_account(
        self.class.configuration_name,
        account_id,
        unlock_account_key_value, # I tried to replace it with get_unlock_account_key and generate_unlock_account_key
) .deliver_later

```
the unlock email doesn't work, and I get 403 Forbidden.
any advice?

**Comment by janko** [comment]
You're probably looking to send the unlock account email right after the account gets locked, not every time Rodauth checks whether the account is locked. It seems this is as easy as:

```rb
after_account_lockout { send_unlock_account_email }
```

**Reply by rida-elbahtouri** [reply]
I want to send the email every time is that possible?

this works

```ruby
after_account_lockout { send_unlock_account_email }
```
but when doing 

```ruby
locked_out? do
    send_unlock_account_email
end
```

I get an email with an empty key

**Reply by janko** [reply]
The following should work:

```rb
# called on first invalid login attempt that resulted in lockout
after_account_lockout { send_unlock_account_email }

# called on subsequent invalid logins
locked_out? do
  locked = super()
  if locked
    @unlock_account_key_value = get_unlock_account_key
    send_unlock_account_email
  end
  locked
end
```

But note that this doesn't handle frequency of sent emails, so this will allow an attacker to DoS your email processing service. You should probably use `unlock_account_request` via the internal_request feature.

**Reply by jeremyevans** [reply]
`send_unlock_account_email` requires the key already be set.  When `after_account_lockout` is called, the key is already set.  Also, you don't want to change the behavior of `locked_out?`, and you only want to send the unlock email if it is locked out, and you shouldn't allow someone to spam unlock account emails.  Maybe:

```ruby
locked_out? do
  if (ret = super()) && !unlock_account_email_recently_sent?
    @unlock_account_key_value = get_unlock_account_key
    set_unlock_account_email_last_sent
    send_unlock_account_email
  end
  ret
end
```

**Reply by rida-elbahtouri** [reply]
I tried both solutions but for some reason, the key doesn't work.
the after_account_lockout email works but I get a 403 Forbidden when I try this

```ruby
 @unlock_account_key_value = get_unlock_account_key
 set_unlock_account_email_last_sent
 send_unlock_account_email
```

**Reply by jeremyevans** [reply]
If you can put together a failing spec or minimal self contained example, I can take a deeper look.

**Reply by rida-elbahtouri** [reply]
I don't have anything custom, here is my rodauth_main code

```ruby
    # number of invalid tries before locking the account
    max_invalid_logins 5
    # sent an other email after 10min
    unlock_account_skip_resend_email_within 600
    
      locked_out? do
      if (ret = super()) && !unlock_account_email_recently_sent?
        @unlock_account_key_value = get_unlock_account_key
        set_unlock_account_email_last_sent
        send_unlock_account_email
      end
      ret
    end
    after_account_lockout do
      set_unlock_account_email_last_sent
      send_unlock_account_email
    end

    create_unlock_account_email do
      RodauthMailer.unlock_account(
        self.class.configuration_name,
        account_id,
        unlock_account_key_value,
      )
    end
```

**Reply by jeremyevans** [reply]
`RodauthMailer` is not something defined by Rodauth.  Maybe this is rodauth-rails specific, in which case you may want to move the discussion to the rodauth-rails repository.

**Reply by rida-elbahtouri** [reply]
I have RodAuth mailer defined
```ruby
class RodauthMailer < ApplicationMailer

 def unlock_account(name, account_id, key)
    @email_link = email_link_key(name, :unlock_account, account_id, key)
    @account = find_account(name, account_id)
    mail to: @account.email, subject: rodauth(name).unlock_account_email_subject
  end

  private

  def email_link_key(name, action, account_id, key)
    instance = rodauth(name)
    instance.instance_variable_set(:@account, { id: account_id })
    instance.instance_variable_set(:"@#{action}_key_value", key)
    instance.public_send(:"#{action}_email_link")
  end
end
```

**Reply by jeremyevans** [reply]
As this involves rodauth-rails, please move this discussion to rodauth-rails.  If you can generate a failing spec or minimal self-contained example for plain Rodauth (without rodauth-rails), then we can continue the discussion here.
